bk-nodes/bk_db.py

```diff
@@ -0,0 +1,241 @@
+       #self.cur.execute(stmt)
+       # stmt = 'LOCK TABLES `nodes_log` READ'
+        self.cur.execute(stmt)
+        #self.db.commit()
+
+        if replay_from_timestamp:
+            # delete all entries newer than replay_from_timestamp in history
+
+            # microseconds in the timestamp lead to wrong behaviour, because microseconds are not in mysql it seems 
+            replay_from_timestamp = replay_from_timestamp.replace(microsecond=0)
+
+            stmt = f'DELETE FROM `nodes_history` WHERE `timestamp` >= %s'
+            print(f'statement: {stmt}', flush=True)
+            try:
+                self.cur.execute(stmt, (replay_from_timestamp,))
+                self.db.commit()
+            except Exception as e:
+                raise Exception("Deleting lastes history failed:" + str(e))
+
+
+        nodes_last_state = {}
+
+        reading_cur = self.db.cursor()
+
+        total_count = None        
+        if replay_from_timestamp:
+            stmt = f'SELECT {fields_str} FROM `nodes_log` WHERE `effective_time` >= %s ORDER BY `effective_time`'
+            #print(f'statement: {stmt}', flush=True)
+            total_count = reading_cur.execute(stmt, (replay_from_timestamp,))
+            debug_stmt = stmt
+            debug_stmt = debug_stmt.replace("%s", f'"{replay_from_timestamp}"', 1)
+            print(f'debug_stmt: {debug_stmt}', flush=True)
+        else: 
+            stmt = f'SELECT {fields_str} FROM `nodes_log` ORDER BY `effective_time`'
+            print(f'statement: {stmt}', flush=True)
+            total_count = reading_cur.execute(stmt)
+
+
+        print(f'total_count: {total_count}')
+        while rows := reading_cur.fetchmany(size=100) :
+           
+
+            print(f'______loop: {len(rows)}', flush=True)
+            for row in rows:
+                print("row: "+str(row), flush=True)
+                
+                #print(rows, flush=True)
+                node_id = row[fieldIndex['node_id']]
+                op = row[fieldIndex['operation']]
+                if op != "insert":
+                    #print(f"ERROR: skipping invalid operation {op}")
+                    raise Exception(f"Operation {op} not supported")
+                    
+                col = row[fieldIndex['field_name']]
+                value = row[fieldIndex['new_value']]
+                effective_time = row[fieldIndex['effective_time']]
+                node_object = nodes_last_state.get(node_id, {})
+                #if not node_object:
+                # TODO get object from db
+                
+                if 'timestamp' in node_object and node_object['timestamp'] != effective_time:
+                    # this is a new timestamp, process old history point first
+                    try:
+                        self.insert_object("nodes_history", node_object)
+                    except Exception as e:
+                        raise Exception("insert_object failed:" + str(e))
```

Comment:
Do we want to wipe the `node_history` database in this condition?

Reply:
Difficult to say, it would depend on the kind of failure. I am less concerned about this step, because the nodes_history is only a temporary cache that can be reconstructed anytime.